New mod Toggle Taskbar Autohide

m1lhaus opened the pull request:
This new mod adds missing feature from 7+ Taskbar Tweaker. Middle clicking on empty taskbar space toggles taskbar autohide feature.

The mod hooks into taskbar event proc callback using code ripped of from [taskbar-volume-control.wh.cpp](https://github.com/ramensoftware/windhawk-mods/blob/main/mods/taskbar-volume-control.wh.cpp). As far as I can tell it is a magic for me, but it works. Whenever a mouse middle click event gets caught, method OnMouseClick() get called. There, using UIAutomation API a name of the element clicked is retrieved. If UIAutomation API returned a taskbar as an element, it means user clicked an empty space on a taskbar. If however other element is returned (start menu, search, etc.), the event is ignored and can be then processed by some other hook.

The reason why I used UIAutomation API is that it reliably returns an identifier of element clicked across Windows version. I am pretty sure this approach will also work for future Windows versions with just small adjustments (e.g. className). Although those are COM calls, since we are dealing here with a mouse click events, I can't see any drawback in the overhead (if there is any).

I tested the mod with Windows 10 22H2, Windows 11 22H1, 22H2, 23H2. I also tested Windows 11 23H2 with Explorer patcher. I tried both 100% DPI and 200% DPI (on WIndows 11 only). I tested clicking both primary and secondary taskbars.

This is my first mod and also the first time ever I did something with Win32 API. So if you see any way for improvement, please let me know.

m417z commented:
Nice. There's some dead code that you can remove, such as (from a short glance): `TaskBarVersionStr`, `SYMBOL_HOOK`, `HookSymbols`, `GetTaskbarViewDllPath`, `case WM_COPYDATA: [...]`.

m417z commented:
BTW looks like the relevant UI Automation definitions were recently added to Wine (and MinGW-w64). I'll look into bundling them with Windhawk's next version.
https://gitlab.winehq.org/wine/wine/-/merge_requests/569

m1lhaus commented:
I removed mentioned dead code. Yeah it would be great to have those UIAutomation bundled since now it makes to code unnecessarily long.

m417z commented on mods/toggle-taskbar-autohide.wh.cpp:
```diff
@@ -0,0 +1,1108 @@
+        return false;
+    }
+
+    IUIAutomationElement *pWindowElement = NULL;
```
FYI you can use COM smart pointers to avoid having to call `Release`. Example:
https://github.com/ramensoftware/windhawk-mods/blob/90d6ccd363cb9764f081ea1981932b2b3dac68cf/mods/taskbar-grouping.wh.cpp#L726-L729

m417z commented on mods/toggle-taskbar-autohide.wh.cpp:
```diff
@@ -0,0 +1,1108 @@
+        return false;
+    }
+
+    BSTR className = NULL;
```
You can use the `_bstr_t` smart pointer here.

m417z commented on mods/toggle-taskbar-autohide.wh.cpp:
```diff
@@ -0,0 +1,1108 @@
+
+    IUIAutomation *pUIAutomation = g_UIAutomation.getInstance();
+    if (pUIAutomation == NULL) {
+        Wh_Log(L"UIAutomationWrapper failed to get instance\n");
```
There's no need to end log lines with a newline.

m417z commented on mods/toggle-taskbar-autohide.wh.cpp:
```diff
@@ -0,0 +1,1108 @@
+    }
+
+    if (!g_UIAutomation.init()) {
+        Wh_Log(L"UIAutomationWrapper failed to initialize\n");
```
It should be before the subclass code, otherwise, the subclass won't be unset and Explorer will crash.

m417z reviewed (commented):
Looks good overall, well done. I left a couple of non-critical comments, let me know if you want to address them, up to you. I can also merge it as is.

I had a similar mod on my list. I wanted to create a replacement to the 7+ Taskbar Tweaker options, which allowed to configure actions for various click events on the taskbar. The events could be e.g. middle click, double click, Ctrl+Alt+click, etc. The actions could be e.g. Show desktop, open start menu, trigger Alt+Tab, run a program, and more. You made the  core, difficult part of it, would you like to extend it and make it customizable? Adding options is fairly easy in Windhawk, I can help and show relevant mod examples.

m1lhaus commented:
I reworked all the findings. Thx for the hint with smart pointers. Man, Microsoft docu is a maze. 

Regarding you point with call order issue in init method. Isn't Wh_ModUninit called every time no matter the result of Wh_ModInit? That would be my expectations.

Regarding old 7+Taskbar functionality, I implemented just the toggle since hiding taskbar is the closest I can get to fullscreen experience on Windows. I didn't really think about the rest. Do you have a code I could re-use? I might take a look into it. If I decide to continue with other features as well, I guess the best would be to rename a mod and to start incrementally implementing those features, right?

m1lhaus commented:
OK, I will look into that and let you know. I'll take down the PR if I decide to give it a shot. Holidays are coming so there should  be some free time for that.

m417z commented:
Another action that you can add that was requested by a user is: Toggle hide desktop icons.
Here's an implementation you can use:
https://gist.github.com/m417z/2ef1273f7998b458c4d4c1002c8f3586

m1lhaus commented:
Thx, implemented. I will finish it soon.

m1lhaus commented:
I implemented most of the stuff from 7+Taskbar tweaker. I skipped Taskbar inspector and Start menu. I can't see any value in the inspector and I experienced crashes with the menu. I might take a look on Start menu feature in future. Now I would like to release what I got so far.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+// ==WindhawkMod==
+// @id              taskbar-empty-space-clicks
+// @name            Click on empty taskbar space
+// @description     Trigger custom action when empty space on a taskbar is double/middle clicked.
```
Descriptions don't usually end with a period.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+
+I will not supporting Insider preview or other minor versions of Windows. However, feel free to [report any issues](https://github.com/m1lhaus/windhawk-mods/issues) related to those versions. I'll appreciate the heads-up in advance.  
+
+In case you are using old Windows taskbar on Windows 11 (**Explorer Patcher** or a similar tool), enable corresponding option on Settings menu. This options will be tested only with the latest major version of Windows 11 (e.g. 23H2). 
```
I recently learned that the precise name is "ExplorerPatcher".

m417z replied:
Still there, but not that important, you can fix it in the next version if you want.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+}
+
+TaskBarAction ParseMouseActionSetting(const wchar_t *option) {
+    const auto value = Wh_GetStringSetting(option);
```
The value is never freed.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+    const GUID XIID_MMDeviceEnumerator = {0xBCDE0395, 0xE52F, 0x467C, {0x8E, 0x3D, 0xC4, 0x57, 0x92, 0x91, 0x69, 0x2E}};
+    if (FAILED(CoCreateInstance(XIID_MMDeviceEnumerator, NULL, CLSCTX_INPROC_SERVER, XIID_IMMDeviceEnumerator, g_pDeviceEnumerator.put_void())) || !g_pDeviceEnumerator) {
+        Wh_Log(L"Failed to create DeviceEnumerator COM instance, ModInit failed");
+        return FALSE;
```
`CoUninitialize` is missing in this case.

m1lhaus replied:
I implemented a wrapper that takes of Uninitialization even if Wh_ModUninit is not called.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+        return false;
+    }
+
+	HWND hChildWnd = FindWindowEx(hParentWnd, NULL, L"SHELLDLL_DefView", NULL);     // parent window of the desktop
```
Some tabs are lurking here.

m417z replied:
Still there, but not that important, you can fix it in the next version if you want.

m417z commented on mods/taskbar-empty-space-clicks.wh.cpp:
```diff
@@ -0,0 +1,1572 @@
+
+// ==WindhawkModSettings==
+/*
+- doubleClickAction: nothing
```
Replace `nothing` with `ACTION_NOTHING`, otherwise the default text just displays "nothing".

![image](https://github.com/ramensoftware/windhawk-mods/assets/4129781/cd80f763-783e-4aaf-80be-4d9de86c6846)

m417z commented:
Looks good overall. I left some small comments that you might want to address, nothing critical.

> I can't see any value in the inspector 

It can actually be handy for operations such as quickly closing multiple windows, and doing other operations such as changing AppIDs which affect grouping on the taskbar. But porting it isn't trivial, I didn't expect that you'd do that.

> I experienced crashes with the menu

Maybe it changed in Windows 11. I'll see if I can find a solution for it.

Also, I think it would be a good idea to add the following to have customization on par with 7+ Taskbar Tweaker:
* Allowing to run arbitrary commands, for example run Notepad/calculator on double-click.
* Adding more buttons (Button 4, button 5, and maybe right click) and modifiers (Ctrl, Alt, Shift). It can be implemented as an array of options, so that uses can add new hotkeys and define an action for each (since listing all permutations is too much). But maybe we can wait and see how many users actually want this.

m417z commented:
Looks good, well done! Merged.

m1lhaus commented:
OK, thx. I was still planning to take a look to StartMenu and to that whole keypress emulation thing as you suggested. I will take care of it in next version then.
